### ISSUE TITLE
Lack of semantics to communicate structure

### ISSUE THREAD

**Issue by daniel-montalvo** [thelounge/thelounge#4311]
Elements such as headings and page regions are missing. They communicate the structure of the page or application to assistive technologies, and help users navigate efficiently to the different sections. Some examples of such regions are:

- List of networks
- List of channels
- Dates in a message list
- New messages

We recommend appropriate use of  headings and page regions to communicate the different sections of the page.

**Comment by brunnre8** [comment]
Would you mind elaborating?
What is missing concretely?

At least for the channel list the structure is annotated

![x](https://labrat.space/irc/be22cc611871d1f3/2021-09-15_15-01-10.png)

**Comment by daniel-montalvo** [comment]
Thanks @brunnre8 for getting back to this so quickly

> Would you mind elaborating?
> What is missing concretely?

I am missing headings that identify the different sections of the page. Screen reader users navigate using specific screen readers functionality that relies on headings.

I would have at least the following:

- h1 with the title of the current network
- h2 before the list of available networks (if the user has several of them available)
- h2 before the list of channels
- h2 before each change in the date of the messages, so that users can jump quickly through the different dates in a long list of messages

For references on how to communicate the structure of a page, see 
https://www.w3.org/WAI/tutorials/page-structure/

> 
> At least for the channel list the structure is annotated
> 
> ![x](https://camo.githubusercontent.com/5b7285c54f899b60fa55ce988a5202b9d3125074475cb6e1d7b697452ff4389b/68747470733a2f2f6c61627261742e73706163652f6972632f626532326363363131383731643166332f323032312d30392d31355f31352d30312d31302e706e67)

I am sorry, I do not understand what you are pointing me too. I am a completely blind person and this seems to me like a screen capture.

From what I am seeing, I have also suggestions for the channels list that I would probably put in a different issue.

**Comment by brunnre8** [comment]
yeah in hindsight that was a stupid thing to do... sorry for that.

the channel list's html contains aria labels and roles which should have been picked up for you, but apparently didn't.

The corresponding elements are divs though

**Comment by daniel-montalvo** [comment]
> yeah in hindsight that was a stupid thing to do... sorry for that.

No worries :-) Just letting you know.
> 
> the channel list's html contains aria labels and roles which should have been picked up for you, but apparently didn't.
> 
> The corresponding elements are divs though

It is really good you have thought of these ARIA labels and roles. There are additional rules for where a specific ARIA role or attribute goes and there are also expectations on what happens to the child elements of a given node when you apply ARIA role. More on this on a separate thread.

**Comment by MaxLeiter** [comment]
Hey @daniel-montalvo, thanks for the detailed issue. 

While we make use of landmarks and ARIA roles, https://usability.yale.edu/web-accessibility/articles/headings is pretty clear we should use headings along with those. 

Regarding "h2 before the list of available networks (if the user has several of them available)": is it acceptable to dynamically change the top level headings like that? I imagine inserting / removing them based on state may cause some confusion. 

Re "h2 before the list of available networks (if the user has several of them available)
h2 before the list of channels", the network list and the channel list are pretty much the same. Do you mean "h2 before the list of channels in the currently open network"?

**Comment by daniel-montalvo** [comment]
> Hey @daniel-montalvo, thanks for the detailed issue.
> 
> While we make use of landmarks and ARIA roles, https://usability.yale.edu/web-accessibility/articles/headings is pretty clear we should use headings along with those.
> 
> Regarding "h2 before the list of available networks (if the user has several of them available)": is it acceptable to dynamically change the top level headings like that? I imagine inserting / removing them based on state may cause some confusion.

It could. A message with and `aria-live` region didicating the currently active network would help avoid that.

> 
> Re "h2 before the list of available networks (if the user has several of them available)
> h2 before the list of channels", the network list and the channel list are pretty much the same. Do you mean "h2 before the list of channels in the currently open network"?

That would be a good solution, given that there is abutton to expand/collapse the network channels.
